locations/spiders/claires.py

```diff
@@ -1,106 +1,27 @@
 # -*- coding: utf-8 -*-
-import json
-import re
-
 import scrapy
 
-from locations.items import GeojsonPointItem
+from locations.brands import Brand
+from locations.seo import extract_ldjson
 from locations.hours import OpeningHours
 
 
-class ClairesSpider(scrapy.Spider):
+class ClairesSpider(scrapy.spiders.SitemapSpider):
     name = "claires"
-    item_attributes = {"brand": "Claire's", "brand_wikidata": "Q2974996"}
+    brand = Brand.from_wikidata("Claire's", "Q2974996")
     allowed_domains = ["claires.com"]
-
-    def start_requests(self):
-        base_url = "https://stores.claires.com/us/"
-
-        states = [
-            "al",
-            "ak",
-            "az",
-            "ar",
-            "ca",
-            "co",
-            "ct",
-            "dc",
-            "de",
-            "fl",
-            "ga",
-            "hi",
-            "id",
-            "il",
-            "in",
-            "ia",
-            "ks",
-            "ky",
-            "la",
-            "me",
-            "md",
-            "ma",
-            "mi",
-            "mn",
-            "ms",
-            "mo",
-            "mt",
-            "ne",
-            "nv",
-            "nh",
-            "nj",
-            "nm",
-            "ny",
-            "nc",
-            "nd",
-            "oh",
-            "ok",
-            "or",
-            "pa",
-            "ri",
-            "sc",
-            "sd",
-            "tn",
-            "tx",
-            "ut",
-            "vt",
-            "va",
-            "wa",
-            "wv",
-            "wi",
-            "wy",
-            "ab",
-            "bc",
-            "mb",
-            "nb",
-            "nl",
-            "ns",
-            "on",
-            "pe",
-            "pr",
-            "qc",
-            "sk",
-            "vi",
-        ]
-
-        for state in states:
-            url = base_url + state
-            yield scrapy.Request(url=url, callback=self.parse_cities)
-
-    def parse_cities(self, response):
-        city_urls = response.xpath(
-            '//*[@class="map-list-item is-single"]/a/@href'
-        ).extract()
-
-        for city_url in city_urls:
-            yield scrapy.Request(url=city_url, callback=self.parse_stores)
-
-    def parse_stores(self, response):
-        store_urls = response.xpath(
-            '//*[@class="map-list-item-header"]/a/@href'
-        ).extract()
-
-        for store_url in store_urls:
-            yield scrapy.Request(url=store_url, callback=self.parse)
+    sitemap_urls = ["https://stores.claires.com/sitemap.xml"]
+    sitemap_rules = [(".html", "parse_store")]
+
+    def parse_store(self, response):
+        for item in extract_ldjson(self.brand, response, "JewelryStore"):
+            if item.has_geo():
```

Comment:
The Game sitemap includes https://storefinder.game.co.uk/game/stores/search which matches my sitemap rules. has ld+json for store type but no geo.
Londis and Premier have a few (fingers of one hand) stores that have no geo. I believe they are to be opened. The number is so small we should not worry.
Pandora has many intermediate pages which match with ld+json but have no geo.
So in all cases I think the spider code is doing what we would want it to do but as said it could be expressed more succintly.

Reply:
Further commit:

So Londis and Premier no longer check for geo.

Pandora spider has a sitemap rule filter that does not pick up "intermediate" pages in store finder structure

Game: keeping approach as it is a lot clearer. For future reference you can have a sitmap rule regex with a target parse function of None, the scrapy framework will ignore such matched URLs.